### DISCUSSION TITLE
[ADD] `eslint-plugin-perfectionist`

### DISCUSSION THREAD

**Discussion by lishaduck** [AndreaPontrandolfo/sheriff#429] [Suggestions for new plugins]
Sorts stuff-keeping stuff consistent is nice, but it also helps performance (very slightly), as objects in a different order aren't optimized by the JIT, and I don't believe any bundlers sort keys, though I'd be happy to be proven wrong.

It could also replace `eslint-simple-import-sort`.

Link: <https://perfectionist.dev/>

I know I can add it manually, but I'd prefer to see if @AndreaPontrandolfo supports the idea, as putting it in sheriff helps provide a better experience for all.
I'd be willing to PR this, likely next week.

**Comment by AndreaPontrandolfo** [comment]
Well, there are a lot of pros and lot of cons.

## PROS
- consistency
- predicatibility
- more stable git diffs

## CONS
- it's a very minor nitpick, people might get annoyed at the warnings (even though most of these probably autofix)
- very high chance to conflict with other rules that mess with ordering with different principles
- performance cost. These rules would have a pretty noticeable impact on editors performance


> but it also helps performance (very slightly), as objects in a different order aren't optimized by the JIT, and I don't believe any bundlers sort keys, though I'd be happy to be proven wrong.

I wouldn't even list this as a PRO, to be honest, these sort of perf optimizations are usually negligible in the real world.

Also has to be said that ESLint has a [core rule](https://eslint.org/docs/latest/rules/sort-keys) for object keys ordering. But it doesn't even autofix, so it's pretty bad.

**Comment by lishaduck** [comment]
> Well, there are a lot of pros and lot of cons.

Yep.


> ## CONS
> * it's a very minor nitpick, people might get annoyed at the warnings (even though most of these probably autofix)
> * very high chance to conflict with other rules that mess with ordering with different principles
>
> ...
> Also has to be said that ESLint has a [core rule](https://eslint.org/docs/latest/rules/sort-keys) for object keys ordering. But it doesn't even autofix, so it's pretty bad.

True enough, but they're getting deprecated in favor of perfectionist.

> * performance cost. These rules would have a pretty noticeable impact on editors performance

Yeah, that's probably worth considering.

> > but it also helps performance (very slightly), as objects in a different order aren't optimized by the JIT, and I don't believe any bundlers sort keys, though I'd be happy to be proven wrong.
> 
> I wouldn't even list this as a PRO, to be honest, these sort of perf optimizations are usually negligible in the real world.

Yeah. I don't think it makes a real difference, but it's not a _negative_.

**Comment by AndreaPontrandolfo** [comment]
> True enough, but they're getting deprecated in favor of perfectionist.

Where did you see this?

**Comment by lishaduck** [comment]
> > True enough, but they're getting deprecated in favor of perfectionist.
> 
> Where did you see this?

https://typescript-eslint.io/rules/sort-type-constituents/, and I thought there was some discussion in the eslint org with the stylistic migration, though a quick search doesn't pop it up.
